pandas/core/algorithms.py

```diff
@@ -911,10 +911,15 @@ def select_n_frame(frame, columns, n, method, keep):
     if not is_list_like(columns):
         columns = [columns]
     columns = list(columns)
+    tmp = Series(frame.index)
+    frame.reset_index(inplace=True, drop=True)
     ser = getattr(frame[columns[0]], method)(n, keep=keep)
     if isinstance(ser, Series):
         ser = ser.to_frame()
```

Comment:
@jreback, i'm a little confused here. I don't see how the current implementation produces the desired result when using multiple columns, for example if we have:
```python
df = pd.DataFrame(dict(
    a=[1, 1, 1, 2, 2],
    b=[4, 3, 8, 9, 5]
))
```

and we call ``df.nsmallest(2, ['a', 'b'])``

There's no code in the current implementation that would try to order by both columns `a` and `b` in that order to give us a resulting frame of
```python
pd.DataFrame(dict(
    a=[1, 1],
    b=[3, 4]
))
```